torch_np/_ndarray.py

```diff
@@ -0,0 +1,520 @@
+            self.tensor, new_shape, refcheck=refcheck
+        )
+
+    nonzero = _funcs.nonzero
+    clip = _funcs.clip
+    repeat = _funcs.repeat
+
+    diagonal = _funcs.diagonal
+    trace = _funcs.trace
+    dot = _funcs.dot
+
+    ### sorting ###
+
+    def sort(self, axis=-1, kind=None, order=None):
+        # ndarray.sort works in-place
+        _funcs.copyto(self, _funcs.sort(self, axis, kind, order))
+
+    argsort = _funcs.argsort
+    searchsorted = _funcs.searchsorted
+
+    ### reductions ###
+    argmax = _funcs.argmax
+    argmin = _funcs.argmin
+
+    any = _funcs.any
+    all = _funcs.all
+    max = _funcs.max
+    min = _funcs.min
+    ptp = _funcs.ptp
+
+    sum = _funcs.sum
+    prod = _funcs.prod
+    mean = _funcs.mean
+    var = _funcs.var
+    std = _funcs.std
+
+    cumsum = _funcs.cumsum
+    cumprod = _funcs.cumprod
+
+    ### indexing ###
+    def item(self, *args):
+        # Mimic NumPy's implementation with three special cases (no arguments,
+        # a flat index and a multi-index):
+        # https://github.com/numpy/numpy/blob/main/numpy/core/src/multiarray/methods.c#L702
+        if args == ():
+            return self.tensor.item()
+        elif len(args) == 1:
+            # int argument
+            return self.ravel()[args[0]]
+        else:
+            return self.__getitem__(args)
+
+    @staticmethod
+    def _upcast_int_indices(index):
+        if isinstance(index, torch.Tensor):
+            if index.dtype in (torch.int8, torch.int16, torch.int32, torch.uint8):
+                return index.to(torch.int64)
+        elif isinstance(index, tuple):
+            return tuple(ndarray._upcast_int_indices(i) for i in index)
+        return index
+
+    def __getitem__(self, index):
+        index = _helpers.ndarrays_to_tensors(index)
+        index = ndarray._upcast_int_indices(index)
+        return ndarray(self.tensor.__getitem__(index))
+
+    def __setitem__(self, index, value):
+        index = _helpers.ndarrays_to_tensors(index)
+        index = ndarray._upcast_int_indices(index)
+        value = _helpers.ndarrays_to_tensors(value)
+        return self.tensor.__setitem__(index, value)
+
+    take = _funcs.take
+
+
+# This is the ideally the only place which talks to ndarray directly.
+# The rest goes through asarray (preferred) or array.
+
+
+def array(obj, dtype=None, *, copy=True, order="K", subok=False, ndmin=0, like=None):
+    _util.subok_not_ok(like, subok)
+    if order != "K":
+        raise NotImplementedError
+
+    # a happy path
+    if isinstance(obj, ndarray):
+        if copy is False and dtype is None and ndmin <= obj.ndim:
+            return obj
+
+    # lists of ndarrays: [1, [2, 3], ndarray(4)] convert to lists of lists
+    if isinstance(obj, (list, tuple)):
+        a1 = []
+        for elem in obj:
+            if isinstance(elem, ndarray):
+                a1.append(elem.tensor.tolist())
+            else:
+                a1.append(elem)
+        obj = a1
```

Comment:
Couldn't we have a list of lists of arrays? In that case this would not work. If we want that to work, this bit needs to be recursive.